signer/src/storage/memory/store.rs

```diff
@@ -0,0 +1,415 @@
+        })
+        .take_while(|stacks_block| {
+            self.bitcoin_blocks
+                .get(&stacks_block.bitcoin_anchor)
+                .is_some_and(|anchor| anchor.block_height >= context_window_end_block.block_height)
+        })
+        .flat_map(|stacks_block| {
+            self.stacks_block_to_withdrawal_requests
+                .get(&stacks_block.block_hash)
+                .cloned()
+                .unwrap_or_default()
+                .into_iter()
+                .map(|pk| {
+                    self.withdrawal_requests
+                        .get(&pk)
+                        .expect("missing withdraw request")
+                        .clone()
+                })
+        })
+        .collect()
+    }
+}
+
+impl Transactable for SharedStore {
+    type Tx<'a> = InMemoryTransaction;
+
+    async fn begin_transaction(&self) -> Result<Self::Tx<'_>, Error> {
+        let store = self.lock().await;
+        let store_clone = store.clone();
+        Ok(InMemoryTransaction {
+            version: store.version,
+            store: Arc::new(Mutex::new(store_clone)),
+            original_store_mutex: Arc::clone(self),
+            completed: AtomicBool::new(false),
+        })
+    }
+}
+
+/// Represents an active in-memory transaction.
+pub struct InMemoryTransaction {
+    /// Records the version of the store at the time of transaction creation.
+    pub version: usize,
+    /// Holds a clone of the store's data for operations within this transaction.
+    pub store: SharedStore,
+    /// Reference to the original store's mutex to commit changes back.
+    pub original_store_mutex: SharedStore,
+    /// Tracks if commit/rollback has been called.
+    pub completed: AtomicBool,
+}
+
+impl TransactionHandle for InMemoryTransaction {
+    async fn commit(self) -> Result<(), Error> {
+        if self.completed.load(std::sync::atomic::Ordering::SeqCst) {
+            panic!("Transaction already completed");
+        }
+
+        // Lock the transaction's clone of the store and get a guard
+        let store = self.store.lock().await.clone();
+        // Lock the original store and get a guard
+        let mut original_store = self.original_store_mutex.lock().await;
+
+        // Naive optimistic concurrency check
+        if self.version != original_store.version {
+            return Err(Error::OptimisticConcurrencyViolation {
+                transaction_version: self.version,
+                store_version: original_store.version,
+            });
+        }
+
+        // Commit the changes from the transactional store to the original store.
+        // This copies all fields from transactional_store_content.
+        *original_store = store.clone();
+
+        // Explicitly set the original store's version to be one greater than the version
+        // this transaction started with. This marks that the store has been updated.
+        original_store.version = self.version + 1;
```

Comment:
IIUC if I start a transaction, do something on the underlying `DbWrite` store, and then `commit`, I will undo everything that happened between the transaction begin and the commit. Am I missing something? Should we increase the `version` on each `DbWrite` operation on the underlying store?

Reply:
Good catch, yeah you're completely right, this is based on the assumption that all writes go through a an `InMemoryTransaction`, but that of course isn't the case.

Reply:
To keep the interfaces consistent (since `Context::get_storage()` and co. return `DbWrite + DbRead + (new) Transactable`), yeah we need some form of "transaction support". The naive optimistic concurrency approach was the simplest I could think of.

Reply:
Got it, then yeah I guess incrementing the version in the writing methods is the "easiest" way.

Reply:
[0d1332a](https://github.com/stacks-sbtc/sbtc/pull/1674/commits/0d1332a647231579b1a691aa0ba08069941800be)

Reply:
Yeah good point, one less thing to remember when writing tests using the in memory db.